x/evm/keeper/context_stack.go

```diff
@@ -0,0 +1,76 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type cachedContext struct {
+	ctx    sdk.Context
+	commit func()
+}
+
+// ContextStack manages the initial context and a stack of cached contexts,
+// to support the `StateDB.Snapshot` and `StateDB.RevertToSnapshot` methods.
+type ContextStack struct {
+	// Context of the initial state before transaction execution.
+	// It's the context used by `StateDB.CommitedState`.
+	initialCtx     sdk.Context
+	cachedContexts []cachedContext
+}
+
+// CurrentContext returns the top context of cached stack,
+// if the stack is empty, returns the initial context.
+func (k *ContextStack) CurrentContext() sdk.Context {
+	l := len(k.cachedContexts)
+	if l == 0 {
+		return k.initialCtx
+	}
+	return k.cachedContexts[l-1].ctx
+}
+
+// Reset sets the initial context and clear the cache context stack.
+func (k *ContextStack) Reset(ctx sdk.Context) {
+	k.initialCtx = ctx
+	if len(k.cachedContexts) > 0 {
+		k.cachedContexts = []cachedContext{}
+	}
+}
+
+// IsClean returns true if the cache context stack is empty.
+func (k *ContextStack) IsClean() bool {
+	return len(k.cachedContexts) == 0
+}
+
+// Commit commits all the cached contexts from top to bottom in order, and clears the stack.
+func (k *ContextStack) Commit() {
+	// commit in order from top to bottom
+	for i := len(k.cachedContexts) - 1; i >= 0; i-- {
+		k.cachedContexts[i].commit()
+	}
+	k.cachedContexts = []cachedContext{}
+}
+
+// Snapshot pushes a new cached context to the stack,
+// and returns the index of it.
+func (k *ContextStack) Snapshot() int {
+	i := len(k.cachedContexts)
+	ctx, commit := k.CurrentContext().CacheContext()
+	k.cachedContexts = append(k.cachedContexts, cachedContext{ctx: ctx, commit: commit})
+	return i
+}
+
+// RevertToSnapshot pops all the cached contexts after(inclusive) the target,
+// panic if the index if out of bound.
+func (k *ContextStack) RevertToSnapshot(target int) {
```

Comment:
my biggest concern with the panic here is calling `RevertToSnapshot(0)` on an empty stack. Should we do a no-op instead in that case?

Reply:
I think the contract on `RevertToSnapshot` could be that it should always be called in pair with `Snapshot`, and there's `RevertAll` that works for the empty stack.